images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go

```diff
@@ -0,0 +1,149 @@
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func (v *VMIPValidator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldVmip, ok := oldObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a old VirtualMachineIP but got a %T", oldObj)
+	}
+
+	newVmip, ok := newObj.(*v1alpha2.VirtualMachineIPAddress)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualMachineIP but got a %T", newObj)
+	}
+
+	v.log.Info("Validate VirtualMachineIP updating", "name", newVmip.Name,
+		"old.type", oldVmip.Spec.Type, "new.type", newVmip.Spec.Type,
+		"old.address", oldVmip.Spec.StaticIP, "new.address", newVmip.Spec.StaticIP,
+	)
+
+	boundCondition, exist := service.GetCondition(vmipcondition.BoundType, oldVmip.Status.Conditions)
+	if exist && boundCondition.Status == metav1.ConditionTrue {
+		if oldVmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeAuto && newVmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
+			v.log.Info("Change the VirtualMachineIP address type to 'Auto' from 'Static' for ip: ", "address", newVmip.Spec.StaticIP)
+			if newVmip.Spec.StaticIP != newVmip.Status.Address {
+				return nil, fmt.Errorf("only type change Auto->Static is allowed: can't change current IP %s to the specified %s", newVmip.Status.Address, newVmip.Spec.StaticIP)
+			}
+
+			err := v.validateSpecFields(newVmip.Spec)
+			if err != nil {
+				return nil, fmt.Errorf("error validating VirtualMachineIP updating: %w", err)
+			}
+
+			return nil, nil
+		}
+
+		if oldVmip.Spec.Type != newVmip.Spec.Type {
+			return nil, errors.New("the VirtualMachineIP is state 'Bound' -> cannot be change type")
+		}
+
+		if newVmip.Spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic && oldVmip.Spec.StaticIP != newVmip.Spec.StaticIP {
+			return nil, errors.New("the VirtualMachineIP is state 'Bound' -> cannot be change static IP")
+		}
+	}
+
+	err := v.validateSpecFields(newVmip.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("error validating VirtualMachineIP updating: %w", err)
+	}
+
+	return nil, nil
+}
+
+func (v *VMIPValidator) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	err := fmt.Errorf("misconfigured webhook rules: delete operation not implemented")
+	v.log.Error(err, "Ensure the correctness of ValidatingWebhookConfiguration")
+	return nil, nil
+}
+
+func (v *VMIPValidator) validateSpecFields(spec v1alpha2.VirtualMachineIPAddressSpec) error {
+	if spec.Type != v1alpha2.VirtualMachineIPAddressTypeStatic && spec.Type != v1alpha2.VirtualMachineIPAddressTypeAuto {
+		return fmt.Errorf("invalid type for VirtualMachineIP: %s. Type must be either 'Static' or 'Auto'", spec.Type)
+	}
+
+	if spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic {
```

Comment:
spec.Type == v1alpha2.VirtualMachineIPAddressTypeStatic && !isValidAddressFormat(spec.StaticIP) {}

or maybe it's better to use the switch-case here, where the default case is fmt.Errorf("invalid type ... ")

Reply:
I think that reworking on the switch may delay the process.

Reply:
Done.